### ISSUE TITLE
hnix-0.9.0 failing to build (broken by repline-0.3.0.0?)

### ISSUE THREAD

**Issue by cdornan** [haskell-nix/hnix#651]
We are seeing the following build failure on stackage nightly which may have been caused by the recent release of `repline-0.3.0.0`. We have suspended hnix from nightly but let us know when the issue has been resolved and we will restore it.

---
```
    Configuring hnix-0.9.0...
    Preprocessing library for hnix-0.9.0..
    Building library for hnix-0.9.0..
    Preprocessing executable 'hnix' for hnix-0.9.0..
    Building executable 'hnix' for hnix-0.9.0..
    [2 of 3] Compiling Repl [System.Console.Repline changed]

    /var/stackage/work/unpack-dir/unpacked/hnix-0.9.0-3049e4182aff661b1e01927a219766f0adee953b95bfa4e67fd33d217a9ea7fb/main/Repl.hs:61:7: error:
        • Couldn't match expected type ‘StateT (IState t0 f0 m0) m ()’
                      with actual type ‘HaskelineT (StateT (IState t2 f2 m3) m3) a0
                                        -> HaskelineT (StateT (IState t2 f2 m3) m3) ExitDecision
                                        -> StateT (IState t2 f2 m3) m3 ()’
        • Probable cause: ‘evalRepl’ is applied to too few arguments
          In the second argument of ‘($)’, namely
            ‘evalRepl
               (return prefix) cmd options (Just ':') completer welcomeText’
          In the expression:
            flip evalStateT initState
              $ evalRepl
                  (return prefix) cmd options (Just ':') completer welcomeText
          In an equation for ‘main’:
              main
                = flip evalStateT initState
                    $ evalRepl
                        (return prefix) cmd options (Just ':') completer welcomeText
                where
                    prefix = "hnix> "
                    welcomeText
                      = liftIO
                          $ putStrLn
                              $ "Welcome to hnix "
                                  <> showVersion version <> ". For help type :help\n"
        • Relevant bindings include
            prefix :: HaskelineT (StateT (IState t2 f2 m3) m3) String
              (bound at main/Repl.hs:63:3)
            main :: m () (bound at main/Repl.hs:60:1)
       |
    61 |     $ evalRepl (return prefix) cmd options (Just ':') completer welcomeText
       |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    /var/stackage/work/unpack-dir/unpacked/hnix-0.9.0-3049e4182aff661b1e01927a219766f0adee953b95bfa4e67fd33d217a9ea7fb/main/Repl.hs:61:36: error:
        • Couldn't match type ‘[Char]’ with ‘Char’
          Expected type: System.Console.Repline.Options
                           (HaskelineT (StateT (IState t2 f2 m3) m3))
            Actual type: [(String, [String] -> Repl e0 t2 f2 m3 ())]
        • In the third argument of ‘evalRepl’, namely ‘options’
          In the second argument of ‘($)’, namely
            ‘evalRepl
               (return prefix) cmd options (Just ':') completer welcomeText’
          In the expression:
            flip evalStateT initState
              $ evalRepl
                  (return prefix) cmd options (Just ':') completer welcomeText
       |
    61 |     $ evalRepl (return prefix) cmd options (Just ':') completer welcomeText
       |                                    ^^^^^^^

    /var/stackage/work/unpack-dir/unpacked/hnix-0.9.0-3049e4182aff661b1e01927a219766f0adee953b95bfa4e67fd33d217a9ea7fb/main/Repl.hs:61:55: error:
        • Couldn't match expected type ‘Maybe String’
                      with actual type ‘CompleterStyle (StateT (IState t1 f1 m1) m1)’
        • In the fifth argument of ‘evalRepl’, namely ‘completer’
          In the second argument of ‘($)’, namely
            ‘evalRepl
               (return prefix) cmd options (Just ':') completer welcomeText’
          In the expression:
            flip evalStateT initState
              $ evalRepl
                  (return prefix) cmd options (Just ':') completer welcomeText
       |
    61 |     $ evalRepl (return prefix) cmd options (Just ':') completer welcomeText
       |                                                       ^^^^^^^^^

    /var/stackage/work/unpack-dir/unpacked/hnix-0.9.0-3049e4182aff661b1e01927a219766f0adee953b95bfa4e67fd33d217a9ea7fb/main/Repl.hs:61:65: error:
        • Couldn't match kind ‘* -> *’ with ‘*’
          When matching types
            m2 :: * -> *
            CompleterStyle :: (* -> *) -> *
          Expected type: CompleterStyle (StateT (IState t2 f2 m3) m3)
            Actual type: m2 ()
        • In the sixth argument of ‘evalRepl’, namely ‘welcomeText’
          In the second argument of ‘($)’, namely
            ‘evalRepl
               (return prefix) cmd options (Just ':') completer welcomeText’
          In the expression:
            flip evalStateT initState
              $ evalRepl
                  (return prefix) cmd options (Just ':') completer welcomeText
        • Relevant bindings include
            welcomeText :: m2 () (bound at main/Repl.hs:64:3)
       |
    61 |     $ evalRepl (return prefix) cmd options (Just ':') completer welcomeText
       |                                                                 ^^^^^^^^^^^
```

**Comment by sorki** [comment]
Thanks for reporting this - I have compared the signatures for `repline 0.2` vs `0.3` and to me it looks like this is due to `repline 0.4` as only that one changes `evalRepl` signature. I'll try fixing this in a compatible manner.

**Comment by sjakobi** [comment]
I have just made a Hackage [revision](http://hackage.haskell.org/package/hnix-0.9.0/revisions/) that adds a `< 0.4` constraint to the `repline` dependency, so users don't unnecessarily run into the build failure.

If you're unfamiliar with Hackage revisions, you can find some info on them here:

https://github.com/haskell-infra/hackage-trustees/blob/master/revisions-information.md

Should I add the same patch to this repo? Alternatively you could just ~~update the~~ add some bounds when you fix the build failure, @sorki.

**Comment by sjakobi** [comment]
Also, @sorki, if backward compatibility turns out to be tricky, I suspect it would be fine to just support `repline >= 0.4.0.0 && < 0.5` for now. We use it only in the executable component.

**Comment by Anton-Latukha** [comment]
Thank you, @cdornan for the report and for your Stackage work.

We just released `0.9.1`, it has `repline >= 0.4.0.0 && < 0.5`.

**Comment by sjakobi** [comment]
I've opened https://github.com/commercialhaskell/stackage/pull/5516 to re-enable `hnix` in Stackage.
